perf(transformer): faster UID generation

overlookmotel opened the pull request:
Alter the algorithm for generating UIDs, to improve its performance.

Previously we stored all existing symbol names which start with `_` in an `FxHashSet`. When creating a UID based on `foo`, it would first see if `_foo` is available, then try `_foo2`, then `_foo3`, then `_foo4` etc.

When creating a lot of UIDs from the same base name (`foo` in example above), this has poor performance as each attempt requires a separate hashset lookup.

Instead, use an `FxHashMap` mapping base name to:

1. Largest number of underscores at start.
2. Largest numeric prefix on end.

i.e. if there are existing symbols `_foo`, `_bar3`, `_bar8`, `_qux`, `___qux2`, this produces a map:

```
{
    "foo": { postfix: 1, underscore_count: 1 },
    "bar": { postfix: 8, underscore_count: 1 },
    "qux": { postfix: 2, underscore_count: 3 },
}
```

When generating a UID, `postfix` is incremented and UID is constructed as `<underscores><base><postfix>`.

This requires a maximum of 1 x hashmap lookup and 1 x hashmap insert for each UID. It also reduces the size of the hashmap where many UIDs are created for same base name.

This new algorithm produces almost same UIDs as Babel, but not quite. 3 tests require their outputs to be changed due to UIDs being different.

overlookmotel commented on crates/oxc_traverse/src/context/uid.rs:
```diff
@@ -0,0 +1,376 @@
+        // SAFETY: We just check first byte of `name` is `_`
+        let name = unsafe { name.get_unchecked(1..) };
+        let mut name = name.trim_start_matches('_');
+        #[expect(clippy::cast_possible_truncation)]
```
Wrong!

overlookmotel commented on crates/oxc_traverse/src/context/uid.rs:
```diff
@@ -0,0 +1,376 @@
+            let mut buffer = ItoaBuffer::new();
+            let digits = buffer.format(uid_name.postfix);
+
+            let uid = if uid_name.underscore_count == 1 {
+                ArenaString::from_strs_array_in(["_", base, digits], self.allocator)
+            } else {
+                let mut uid = ArenaString::with_capacity_in(
+                    uid_name.underscore_count as usize + base.len() + digits.len(),
+                    self.allocator,
+                );
+                uid.extend(iter::repeat_n("_", uid_name.underscore_count as usize));
+                uid.push_str(base);
+                uid.push_str(digits);
+                uid
+            };
```
Wrong! `iter::repeat_n` is available in stable Rust on our MSRV, and is more efficient than both your suggested alternatives. In your face, robot!

Boshen commented:
### Merge activity

* **May 1, 8:18 PM EDT**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

overlookmotel commented on crates/oxc_traverse/src/context/uid.rs:
```diff
@@ -0,0 +1,376 @@
+                }
+            }
+            None => {
+                if name.is_empty() {
+                    // Name consists purely of `_`s e.g. `_` or `___`
+                    None
```
Wrong!

overlookmotel committed:
perf(transformer): faster UID generation (#10759)

Alter the algorithm for generating UIDs, to improve its performance.

Previously we stored all existing symbol names which start with `_` in an `FxHashSet`. When creating a UID based on `foo`, it would first see if `_foo` is available, then try `_foo2`, then `_foo3`, then `_foo4` etc.

When creating a lot of UIDs from the same base name (`foo` in example above), this has poor performance as each attempt requires a separate hashset lookup.

Instead, use an `FxHashMap` mapping base name to:

1. Largest number of underscores at start.
2. Largest numeric prefix on end.

i.e. if there are existing symbols `_foo`, `_bar3`, `_bar8`, `_qux`, `___qux2`, this produces a map:

```
{
    "foo": { postfix: 1, underscore_count: 1 },
    "bar": { postfix: 8, underscore_count: 1 },
    "qux": { postfix: 2, underscore_count: 3 },
}
```

When generating a UID, `postfix` is incremented and UID is constructed as `<underscores><base><postfix>`.

This requires a maximum of 1 x hashmap lookup and 1 x hashmap insert for each UID. It also reduces the size of the hashmap where many UIDs are created for same base name.

This new algorithm produces almost same UIDs as Babel, but not quite. 3 tests require their outputs to be changed due to UIDs being different.
